parley/src/layout/line/greedy.rs

```diff
@@ -51,196 +117,248 @@ impl<'a, B: Brush> BreakLines<'a, B> {
+                            if is_space {
+                                self.state.line.num_spaces += 1;
+                            }
+                        }
+                        // Else we line break:
+                        else {
+                            // Handle case where cluster is space character. Hang overflowing whitespace.
+                            if is_space {
+                                self.state.append_cluster_to_line(next_x);
+                                if try_commit_line!(BreakReason::Regular) {
+                                    // TODO: can this be hoisted out of the conditional?
+                                    self.state.cluster_idx += 1;
+                                    return self.start_new_line();
+                                }
+                            }
+                            // Handle the (common) case where we have previously encountered a line-breaking opportunity in the current line
+                            //
+                            // We "take" the line-breaking opportunity by starting a new line and resetting our
+                            // item/run/cluster iteration state back to how it was when the line-breaking opportunity was encountered
+                            else if let Some(prev) = self.state.prev_boundary.take() {
+                                println!("REVERT");
+                                debug_assert!(prev.state.x != 0.0);
+
+                                // Q: Why do we revert the line state here, but only revert the indexes if the commit suceeds?
+                                self.state.line = prev.state;
+                                if try_commit_line!(BreakReason::Regular) {
+                                    // Revert boundary state to prev state
+                                    self.state.item_idx = prev.item_idx;
+                                    self.state.run_idx = prev.run_idx;
+                                    self.state.cluster_idx = prev.cluster_idx;
+
+                                    return self.start_new_line();
+                                }
+                            }
+                            // Otherwise perform an emergency line break
+                            // TODO: make emergency line breaks controllable via a setting
+                            else {
+                                // If we're at the start of the line, this particular cluster will never fit,
+                                // so consume it and accept the overflow.
+                                if self.state.line.x == 0. {
+                                    self.state.append_cluster_to_line(next_x);
+                                    self.state.cluster_idx += 1;
+                                }
+                                if try_commit_line!(BreakReason::Emergency) {
+                                    self.state.cluster_idx += 1;
```

Comment:
I've not tested it, but it seems like this would always skip a cluster? This doesn't look the same as other code above (but be aware that I'm only doing a single-pass review on unfamiliar code)

Reply:
I would also like to understand how/why/if this works correctly (and I don't). But as Chad says, this code is unchanged from the existing code so I don't think it should be blocking.